content/guest-post/session-based-authentication-with-nestjs-and-mongodb/index.md

```diff
@@ -0,0 +1,977 @@
+
+`local.auth.guard.ts`
+
+```js
+    import { ExecutionContext, Injectable } from '@nestjs/common';
+    import { AuthGuard } from '@nestjs/passport';
+    @Injectable()
+    export class LocalAuthGuard extends AuthGuard('local') {
+      async canActivate(context: ExecutionContext) {
+        const result = (await super.canActivate(context)) as boolean;
+        const request = context.switchToHttp().getRequest();
+        await super.logIn(request);
+        return result;
+      }
+    }
+```
+
+If you log in now, you would see the session ID stored in a cookie, which is just a key to the session store, and the cookie gets saved in the browser. The cookie is automatically attached to the rest of the request.
+
+Now that the session is working, you can access the protected route, it would return the expected user’s details.
+
+## Logout Users
+
+As mentioned earlier, once a user logs out, you destroy all sessions.
+
+To log out a user, go to the `users.controller.ts` file, add a logout route, and call the `req.session.session()` method. You can return a message notifying that the user’s session has ended.
+
+```js
+    import {
+      Body,
+      Controller,
+      Get,
+      Post,
+      UseGuards,
+      Request,
+    } from '@nestjs/common';
+    import * as bcrypt from 'bcrypt';
+    import { AuthenticatedGuard } from 'src/auth/authenticated.guard';
+    import { LocalAuthGuard } from 'src/auth/local.auth.guard';
+    import { UsersService } from './users.service';
+    @Controller('users')
+    export class UsersController {
+      constructor(private readonly usersService: UsersService) {}
+      //signup
+      @Post('/signup')
+      async addUser(
+        @Body('password') userPassword: string,
+        @Body('username') userName: string,
+      ) {
+        const saltOrRounds = 10;
+        const hashedPassword = await bcrypt.hash(userPassword, saltOrRounds);
+        const result = await this.usersService.insertUser(
+          userName,
+          hashedPassword,
+        );
+        return {
+          msg: 'User successfully registered',
+          userId: result.id,
+          userName: result.username
+        };
+      }
+      //Post / Login
+      @UseGuards(LocalAuthGuard)
+      @Post('/login')
+      login(@Request() req): any {
+        return {User: req.user,
+                msg: 'User logged in'};
+      }
+       //Get / protected
+      @UseGuards(AuthenticatedGuard)
+      @Get('/protected')
+      getHello(@Request() req): string {
+        return req.user;
+      }
+       //Get / logout
+      @Get('/logout')
+        logout(@Request() req): any {
+          req.session.destroy();
+          return { msg: 'The user session has ended' }
+        }
+    }
+```
+
+So, once you log out, it returns a message notifying you that the user session has ended. The code for this tutorial is hosted [here](https://github.com/icode247/Session-based-Auth-with-Nestjs-MongoDB-) on my Github repository.
```

Comment:
@icode247 Please add the example/sample code in the following repo by raising a separate pull request: https://github.com/LoginRadius/engineering-blog-samples